### ISSUE TITLE
`single_match` shouldn't warn when matching on an explicit error type

### ISSUE THREAD

**Issue by stevenengler** [rust-lang/rust-clippy#13371]
### Summary

In the following code, I think the clippy suggestion is worse than the original code. The original code is explicitly matching on the error "NotFoundError". If the error return type of `foo` were to change in the future (for example to become an enum with multiple error variants, or to become something like "NotAuthenticatedError" instead), we'd want this match statement to fail to build so that we could update it for the new error type and make sure we're handling it correctly. The clippy suggestion wants us to remove the error matching completely.

I think it would be nicer if this lint only triggered when the error type is omitted completely, for example `Err(_) => {}`.

### Lint Name

single_match

### Reproducer

I tried this code:

```rust
struct NotFoundError;

fn bar() -> Result<(), NotFoundError> { todo!() }

fn main() {
    match bar() {
        Ok(()) => {},
        // don't need to do anything if not found
        Err(NotFoundError) => {},
    }
}
```

I saw this happen:

```text
warning: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
  --> src/main.rs:6:5
   |
6  | /     match bar() {
7  | |         Ok(()) => {},
8  | |         // don't need to do anything if not found
9  | |         Err(NotFoundError) => {},
10 | |     }
   | |_____^ help: try: `if let Ok(()) = bar() {}`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_match
   = note: `#[warn(clippy::single_match)]` on by default
```

I expected to see this happen:

No warning.

### Version

```text
rustc 1.83.0-nightly (26b5599e4 2024-09-06)
binary: rustc
commit-hash: 26b5599e4d6ed2b45152c60493c1788c0a27533d
commit-date: 2024-09-06
host: x86_64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.0
```

**Comment by ia0** [comment]
This is more general than errors. I have the same false positive when matching on `Some(())` to make sure that I have an `Option<()>` and not some arbitrary `Option<T>`. Here's the code:

```rust
match foo() {
    None => process(),
    Some(()) => (), // makes sure foo() returns Option<()> and we don't lose information
}
```

Clippy suggests to use `if foo() == None { process() }` which could introduce issues in future refactorings if `foo()` starts to return possibly useful information in the `Some` case.

Clippy should only trigger if the `Some` branch uses a catch-all like `Some(_)` which explicitly discards information.
